yellowbox/extras/websocket.py

```diff
@@ -0,0 +1,356 @@
+                # Not supposed to happen.
+                logger.warning("Message received with no path")
+                return
+
+            assert self._get_generator
+            get_generator = self._get_generator()  # Resolve weakref
+            assert get_generator
+            generator_function = get_generator(path)
+
+            if generator_function is None:
+                logger.info(f"No handler assigned to {path}. "
+                            "Closing connection.")
+                return
+
+            self._generator = generator_function(self)
+            initialized = True
+        except Exception:
+            logger.exception("Exception raised on generator start.")
+            raise
+
+        finally:
+            if not initialized:
+                self.close()
+
+        self._advance_generator(None)
+
+    def _advance_generator(
+            self, data: Union[bytearray, str, None]) -> None:
+        try:
+            try:
+                # First one will send None, rest will send data.
+                msg = self._generator.send(data)  # type: ignore
+            except StopIteration as exc:
+                if exc.value:
+                    self.send_message(exc.value)  # May throw an exception.
+                self.close()
+            else:
+                if msg is not None:
+                    self.send_message(msg)
+        except Exception:
+            logger.exception("Exception raised while handling callback.")
+            raise  # Handle_close will be called
+
+    def handle(self) -> None:
+        self._advance_generator(self.data)
+
+    def handle_close(self) -> None:
+        err = ConnectionAbortedError()
+        generator = self._generator
+        if generator is None:
+            return
+        del self._generator
+
+        try:
+            generator.throw(err)
+        except Exception as exc:
+            # Check if it's our exception. If it is, ignore it.
+            if exc is err:
+                return
+            logger.exception("Exception raised on connection close.")
+            raise
+        finally:
+            generator.close()  # May throw an exception. Ignore it.
+
+
+@no_type_check
+def _to_generator(side_effect: SIDE_EFFECT_TYPE
+                  ) -> _GEN_FUNCTION_TYPE:
+    """Convert a side effect to a generator function.
+
+    Args:
+        side_effect: See WebsocketService.route().
+
+    Returns:
+        Generator function, same as the one in WebsocketService.route().
+    """
+    # Side effect == normal string
+    if isinstance(side_effect, (str, bytes, bytearray, memoryview)):
+        def gen(*args: Any, **kwargs: Any) -> _GENERAOTR_TYPE:
+            return side_effect
+            yield  # On purpose.
+        return gen
+    
+    # Side effect == list of strings
+    if isinstance(side_effect, Iterable):
+        def gen(*args, **kwargs) -> _GENERAOTR_TYPE:
+            for item in side_effect:
+                yield item
```

Comment:
be nice to fellow generators
```suggestion
            return (yield from side_effect)
```

Reply:
I would, if it wouldn't cause bugs